homeassistant/components/sensor/__init__.py

```diff
@@ -273,6 +277,67 @@ def state(self) -> Any:
         """Return the state of the sensor and perform unit conversions, if needed."""
         unit_of_measurement = self.native_unit_of_measurement
         value = self.native_value
+        device_class = self.device_class
+
+        # We have an old non-datetime value, warn about it and convert it during
+        # the deprecation period.
+        if (
+            value is not None
+            and device_class in (DEVICE_CLASS_DATE, DEVICE_CLASS_TIMESTAMP)
+            and not isinstance(value, (date, datetime))
+        ):
+            # Deprecation warning for date/timestamp device classes
+            if not self.__datetime_as_string_deprecation_logged:
+                caller = inspect.stack()[2]
```

Comment:
Can we use `self._suggest_report_issue`?

Reply:
uh... dunno what that is, will take a look 👀

Reply:
Done.